### PR TITLE
Remove various state.Model constructors

### PR DESCRIPTION
**Pull request by mjs**
## Description of change

This change makes most of the steps so that `state.Model` instances can only be created from `state.State` associated to that model. There were many ways to create a `state.Model` from *any* `State` which meant that the `Model` referred to a `state.State` associated with the wrong model. This is dangerous - requiring great care - and limits what can be done by `Model` instances.

Much functionality will soon be moved onto `state.Model`, requiring that the `state.State` it references is the correct one.

Change highlights:

- `State.ControllerModel()` has been replaced by `State.ControllerModelTag()` and `State.ControllerModelUUID()`
- `State.AllModels()` has been replaced by `State.AllModelUUIDs()`
- `State.ModelsForUser()` now returns UUIDs instead of `UserModels`. `state.UserModel` no longer exists.
- Fixed a bug in `State.ModelsForUser()` where dead or importing models could be reported for admin users.
- Access to `Model` instances is now done via `state.State` in most cases. Added a `StatePool.GetModel()` helper as a convenience.
- Many changes in the apiserver package resulted as fallout from the above. This resulted in most calls to `State.ForModel()` being removed (bug 1698702)

There is still one `state.Model` constructor to remove: `state.State.GetModel()`. This will be done in the next PR - this one was getting big enough.

## QA steps

Confirmed that core model, controller and migration functionality worked:

```
$ juju bootstrap lxd source
# wait
$ juju bootstrap lxd target
# wait
$ juju switch source
$ juju add-model foo
$ juju deploy ubuntu
# wait
$ juju migrate foo target
# wait
$ juju debug-log -m source:controller -l ERROR # check controller logs for errors
$ juju debug-log -m target:controller -l ERROR # check controller logs for errors
$ juju destroy-controller -y --destroy-all-models source
$ juju destroy-controller -y --destroy-all-models target
```

## Documentation changes

N.A. - internal only

## Bug reference

Drive-by changes fixed https://bugs.launchpad.net/juju/+bug/1698702

### CODE REVIEW

File: apiserver/common/modelmanagerinterface.go [hunk 1]
```diff
@@ -99,21 +101,13 @@ var _ ModelManagerBackend = (*modelManagerStateShim)(nil)
 
 type modelManagerStateShim struct {
 	*state.State
+	pool *state.StatePool
```
Comment (wallyworld) [inline apiserver/common/modelmanagerinterface.go]: Not sure i agree with this - the state shim is supposed to only wrap state to provide access to its methods.
I rather stick to keeping state and state pool separate in these shim objects to reflect the semantics of those structs

File: apiserver/facades/client/applicationoffers/applicationoffers.go [hunk 2]
```diff
@@ -91,7 +91,7 @@ func NewOffersAPI(ctx facade.Context) (*OffersAPI, error) {
 	return createOffersAPI(
 		GetApplicationOffers,
 		environFromModel,
-		GetStateAccess(ctx.State()),
+		GetStateAccess(ctx.State(), ctx.StatePool()),
```
Comment (wallyworld) [inline apiserver/facades/client/applicationoffers/applicationoffers.go]: we have a state pool below, prefer to keep them seperate

File: apiserver/common/modelmanagerinterface.go (hunk 1 above)
Comment (mjs) [inline apiserver/common/modelmanagerinterface.go]: I had them separate initially but it worked out much nicer to wrap them together. The backend interface then just represents everything to do with "state". It make the tests cleaner too.

File: apiserver/facades/client/applicationoffers/applicationoffers.go (hunk 2 above)
Comment (mjs) [inline apiserver/facades/client/applicationoffers/applicationoffers.go]: I had missed that a pool was being used separately here. I would actually like to update the PR to extend the Backend interface to have a GetBackend method and not pass the pool directly to createOffersAPI at all.

Happy to discuss further of course.

File: state/modeluser.go [hunk 3]
```diff
@@ -170,85 +170,62 @@ func (st *State) removeModelUser(user names.UserTag) error {
...
-		return nil, errors.Trace(err)
-	}
-	var result []*UserModel
-	for _, model := range models {
-		result = append(result, &UserModel{Model: model, User: model.Owner()})
-	}
-	return result, nil
-}
-
 // ModelsForUser returns a list of models that the user
```
Comment (wallyworld) [inline state/modeluser.go]: It really returns the UUIDs right?
SHould these methods that still say "Models" be renamed to "ModelUUIDs" for better clarity?

File: apiserver/facades/client/applicationoffers/applicationoffers.go (hunk 2 above)
Comment (wallyworld) [inline apiserver/facades/client/applicationoffers/applicationoffers.go]: as in GetBackend(uuid)? Like is done for model manager backend I'm guessing

File: state/modeluser.go (hunk 3 above)
Comment (mjs) [inline state/modeluser.go]: Done

**Comment by mjs** [comment]
$$merge$$

**Pull request merged**